dask_expr/merge.py

```diff
@@ -0,0 +1,226 @@
+        "right_index": False,
+        "suffixes": ("_x", "_y"),
+        "indicator": False,
+        "backend": None,
+        "shuffle_backend": None,
+    }
+
+    def __str__(self):
+        return f"Merge({self._name[-7:]})"
+
+    @property
+    def _kwargs(self):
+        return {
+            k: self.operand(k)
+            for k in [
+                "how",
+                "on",
+                "left_on",
+                "right_on",
+                "left_index",
+                "right_index",
+                "suffixes",
+                "indicator",
+            ]
+        }
+
+    @functools.cached_property
+    def _meta(self):
+        left = self._as_meta(self.left)
+        right = self._as_meta(self.right)
+        return left.merge(right, **self._kwargs)
+
+    def _divisions(self):
+        npartitions_left = self._as_expr(self.left).npartitions
+        npartitions_right = self._as_expr(self.right).npartitions
+        npartitions = max(npartitions_left, npartitions_right)
+        return (None,) * (npartitions + 1)
+
+    @staticmethod
+    def _as_meta(dep, nonempty=True):
+        if isinstance(dep, Expr):
+            dep = dep._meta
+        return meta_nonempty(dep) if nonempty else make_meta(dep)
+
+    @staticmethod
+    def _as_expr(dep):
+        if isinstance(dep, Expr):
+            return dep
+        return FromPandas(dep, 1)
+
+    def _simplify_down(self):
+        # Lower from an abstract expression using
+        # logic in MergeBackend.from_abstract_merge
+        backend = self.backend or MergeBackend
+        if hasattr(backend, "from_abstract_merge"):
+            return backend.from_abstract_merge(self)
+        else:
+            raise ValueError(f"{backend} not supported")
+
+
+class MergeBackend(Merge):
+    """Base merge-backend class"""
+
+    def _simplify_down(self):
+        return None
+
+    @classmethod
+    def from_abstract_merge(cls, expr: Merge) -> Expr:
+        """Return a new Exprs to perform a merge"""
+
+        # TODO:
+        #  1. Handle mixed indexed merge
+        #  2. Add multi-partition broadcast merge
+        #  3. Add/leverage partition statistics
+
+        left = expr._as_expr(expr.left)
+        right = expr._as_expr(expr.right)
+        how = expr.how
+        on = expr.on
+        left_on = expr.left_on
+        right_on = expr.right_on
+        left_index = expr.left_index
+        right_index = expr.right_index
+
+        for o in [on, left_on, right_on]:
+            if isinstance(o, Expr):
+                raise NotImplementedError()
+        if (
+            not on
+            and not left_on
+            and not right_on
+            and not left_index
+            and not right_index
+        ):
+            on = [c for c in left.columns if c in right.columns]
+            if not on:
+                left_index = right_index = True
+
+        if on and not left_on and not right_on:
+            left_on = right_on = on
+            on = None
```

Comment:
I would hope that this stuff could happen in the user API before we ever got to expressions.  Is that possible?

Reply:
Yes - In this case we can.